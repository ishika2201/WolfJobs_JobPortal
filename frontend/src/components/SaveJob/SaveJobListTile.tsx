import { useEffect, useState } from "react";
import { useSearchParams, useNavigate } from "react-router-dom"; // Import useNavigate
import { useApplicationStore } from "../../store/ApplicationStore";
import { useUserStore } from "../../store/UserStore";

const SaveJobListTile = ({ data }: { data: Job }) => {
  const navigate = useNavigate(); // Initialize useNavigate
  const [active, setActive] = useState<boolean>(false);
  const [application, setApplication] = useState<Application | null>(null);
  const [searchParams, setSearchParams] = useSearchParams();
  const userId = useUserStore((state) => state.id);
  const userRole = useUserStore((state) => state.role);
  const applicationList: Application[] = useApplicationStore((state) => state.applicationList);

  useEffect(() => {
    const temp: Application | undefined = applicationList.find(
      (item: Application) => item.jobid === data._id && item.applicantid === userId
    );
    setApplication(temp || null);
  }, [data, applicationList, userId]);

  const affilation = data.managerAffilication;
  const role = data.name;
  const jobType = data?.type?.split("-")?.join(" ");
  const pay = data.pay || "0";

  useEffect(() => {
    const id = searchParams.get("jobId");
    setActive(data._id === id);
  }, [searchParams, data]);

  const handleClick = (e: React.MouseEvent) => {
    e.preventDefault();
    navigate(`/explore?jobId=${data._id}`);
  };

  const getAffiliationTag = (tag: string) => {
    return tag.split("-").join(" ");
  };

  const getAffiliationColour = (tag: string) => {
    if (tag === "nc-state-dining") {
      return "bg-[#FF2A2A]/10";
    } else if (tag === "campus-enterprises") {
      return "bg-[#91B0FF]/10";
    } else if (tag === "wolfpack-outfitters") {
      return "bg-[#FBD71E]/10";
    }
    return "bg-[#FF2A2A]/10";
  };

  return (
    <div className="my-3" onClick={handleClick}>
      <div className={`p-3 bg-white rounded-xl shadow-sm ${active ? "border-black" : "border-white"} border`}>
        <div className="flex flex-row">
          <div className="w-4/6 ">
            <div className={`w-fit ${getAffiliationColour(affilation)} rounded-2xl px-3 py-0`}>
              <p className="inline text-xs" style={{ width: "fit-content" }}>
                {getAffiliationTag(affilation).toUpperCase()}
              </p>
            </div>
            <div className="h-1"></div>
            <div className="pl-2">
              <p className="text-base">
                <b>Role:</b> {role}
              </p>
              <p className="text-base">
                <b>Job Status:</b>
                <span className={`${data.status === "closed" ? "text-[#FF5353]" : ""}`}>
                  &nbsp;<span className="capitalize">{data.status}</span>
                </span>
              </p>
              <p className="text-base">
                <b>Type:</b> <span className="capitalize"> {jobType} </span>
              </p>
              {userRole === "Applicant" && (
                <p className="text-base">
                  <b>Application Status:</b>&nbsp;
                  {application ? <span className="capitalize">{application?.status}</span> : "In Review"}
                </p>
              )}
            </div>
          </div>
          <div className="w-2/6 flex flex-col-reverse text-right">
            <p className="text-3xl">{pay}$/hr</p>
          </div>
        </div>
      </div>
    </div>
  );
};

export default SaveJobListTile;
