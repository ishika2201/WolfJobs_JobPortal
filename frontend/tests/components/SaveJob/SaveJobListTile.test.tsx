import { render, screen } from "@testing-library/react";
import SaveJobListTile from "../../../src/components/SaveJob/SaveJobListTile";
import { MemoryRouter } from "react-router-dom";
import { describe, it, expect } from "vitest";

describe("SaveJobListTile", () => {
  const jobData = {
    _id: "1",
    managerAffilication: "nc-state-dining",
    pay: "100",
    status: "closed",
    name: "Software Developer",
    type: "full-time",
  };

  it("renders SaveJobListTile with job details", () => {
    render(
      <MemoryRouter>
        <SaveJobListTile data={jobData} />
      </MemoryRouter>
    );
    expect(screen.getByText(/nc state dining/i)).toBeInTheDocument();
    expect(screen.getByText(/software developer/i)).toBeInTheDocument();
    expect(screen.getByText(/closed/i)).toBeInTheDocument();
    expect(screen.getByText(/full time/i)).toBeInTheDocument();
    expect(screen.getByText("100$/hr")).toBeInTheDocument();
  });

  it("displays the job status as open with no color class if the job is open", () => {
    render(
      <MemoryRouter>
        <SaveJobListTile data={{ ...jobData, status: "open" }} />
      </MemoryRouter>
    );
    const status = screen.getByText(/open/i);
    expect(status).not.toHaveClass("text-[#FF5353]");
  });

  it("does not render pay section if no pay information is provided", () => {
    render(
      <MemoryRouter>
        <SaveJobListTile data={{ ...jobData, pay: "" }} />
      </MemoryRouter>
    );
    expect(screen.queryByText(/100\$/)).not.toBeInTheDocument();
  });

  it("renders 'part-time' if the job type is part-time", () => {
    render(
      <MemoryRouter>
        <SaveJobListTile data={{ ...jobData, type: "part-time" }} />
      </MemoryRouter>
    );
    expect(screen.getByText(/part time/i)).toBeInTheDocument();
  });

  it("renders 'contract' if the job type is contract", () => {
    render(
      <MemoryRouter>
        <SaveJobListTile data={{ ...jobData, type: "contract" }} />
      </MemoryRouter>
    );
    expect(screen.getByText(/contract/i)).toBeInTheDocument();
  });
});
