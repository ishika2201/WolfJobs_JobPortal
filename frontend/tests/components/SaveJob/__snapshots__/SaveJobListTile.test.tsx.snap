// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SaveJobListTile > matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="my-3"
  >
    <div
      class="p-3 bg-white rounded-xl shadow-sm border-white border"
    >
      <div
        class="flex flex-row"
      >
        <div
          class="w-4/6 "
        >
          <div
            class="w-fit bg-[#FF2A2A]/10 rounded-2xl px-3 py-0"
          >
            <p
              class="inline text-xs"
            >
              NC STATE DINING
            </p>
          </div>
          <div
            class="h-1"
          />
          <div
            class="pl-2"
          >
            <p
              class="text-base"
            >
              <b>
                Role:
              </b>
               Software Developer
            </p>
            <p
              class="text-base"
            >
              <b>
                Job Status:
              </b>
              <span
                class="text-[#FF5353]"
              >
                Â 
                <span
                  class="capitalize"
                >
                  closed
                </span>
              </span>
            </p>
            <p
              class="text-base"
            >
              <b>
                Type:
              </b>
               
              <span
                class="capitalize"
              >
                 full time 
              </span>
            </p>
          </div>
        </div>
        <div
          class="w-2/6 flex flex-col-reverse text-right"
        >
          <p
            class="text-3xl"
          >
            100$/hr
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
